packages/aztec.js/src/proof/dividend/index.js

```diff
@@ -0,0 +1,138 @@
+const { constants, errors, proofs } = require('@aztec/dev-utils');
+const BN = require('bn.js');
+const { AbiCoder } = require('web3-eth-abi');
+const { keccak256, padLeft } = require('web3-utils');
+
+const bn128 = require('../../bn128');
+const { inputCoder, outputCoder } = require('../../encoder');
+const { Proof, ProofType } = require('../proof');
+const ProofUtils = require('../utils');
+
+const { AztecError } = errors;
+const { BN128_GROUP_REDUCTION } = constants;
+
+class DividendProof extends Proof {
+    constructor(notionalNote, residualNote, targetNote, sender, za, zb, metadata = [residualNote, targetNote]) {
```

Comment:
We should probably add comments describing what these input parameters are.  

`notionalNote` : the note that ones is computing a dividend of.  
`residualNote` : the dividend computation will have an integer rounding error, this is represented by the residual note (as having a public remainder would leak information)  
`targetNote`: the note that is being produced  
`za/zb` : public integers that define a ratio between notionalNote and targetNote.

It is also probably imporant to document that this proof assumes that it is in the prover's interest for `targetNote` to be as large as possible. What we are actually checking, is that `targetNote <= (zb / za) * notionalNote. So a prover can, if they desire, create a `targetNote` that is smaller than the note they are entitled to.  

If one wishes to use this proof to compute a debt, the roles of `targetNote` and `notionalNote` need to be swapped around, so that it is still in the prover's interest to minimize `targetNote`.